### PR TITLE
Modelling date/times based on java.time types

### PR DESCRIPTION
**Pull request by idc101**
Also ignored the entire `.idea` directory.

### CODE REVIEW

File: src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala [hunk 1]
```diff
@@ -53,8 +53,8 @@ trait LocalAnalyticsModule extends AnalyticsModule {
   implicit override val byteType: Type[scala.Byte]            = LocalType(Reified.Byte)
   implicit override val nullType: Type[scala.Null]            = LocalType(Reified.Null)
   implicit override val shortType: Type[scala.Short]          = LocalType(Reified.Short)
-  implicit override val timestampType: Type[LocalDateTime]    = LocalType(Reified.Timestamp)
-  implicit override val dateType: Type[LocalDate]             = LocalType(Reified.Date)
+  implicit override val instantType: Type[Instant]            = LocalType(Reified.Instant)
```
Comment (camjo) [inline src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala:56]: The more I think about this the more I think we might need a few more types here... Even between these two types here:
- `Instant` is a timestamp with a timezone (offset from epoch)
- `LocalDateTime` is a timestamp without a timezone
Both seem to have plenty of valid use cases.

This is actually one of the consequences of controlling the language users can define - we have to reimplement a *lot* of functionality. In this case seemingly `java.time.*`.

@jdegoes How did you envisage this looking?

**Review by camjo — COMMENTED** [review]
You need to run `sbt fmt` or else Travis won't build!

File: src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala (hunk 1 above)
Comment (jdegoes) [reply src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala:56]: Our capacity to do fancy time manipulations will be quite heavily limited in some environments (e.g. GPU), in which we don't necessarily have access to the equivalent of a time zone database. So I suggest we focus on the smallest set of primitives and let the user do most of the work (or we can do it later, by building on top the core primitives). 

[This](https://www.cloudera.com/documentation/enterprise/5-6-x/topics/impala_timestamp.html) seems a reasonable place to start. Basically with a timezone-less format, and extraction methods (for day, month, year, hour, minute, etc), the user can do quite a lot on their own (including storing timezone in a separate field). And we can support such functionality everywhere in a very performant fashion.

Comment (ritschwumm) [reply src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala:56]: managing an offset from the epoch (like Instant does) is simple and gets you quite far. as soon as you are starting to think about things like timezones or calendars, you're in deep shit - don't go there if you can avoid it.

even LocalDateTime is tricky - who said _your_ country used the gregorian calendar at all at that date, for example?

Comment (idc101) [reply src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala:56]: Yes so using `Instant` is the same strategy as the Impala Timestamp. Just storing an offset with no timezone, and interpreting it relative to UTC. You would be free to store the Timezone separately.

@ritschwumm agree about `LocalDate` but I'd argue the general usefulness outweighs that as an edge case.

Comment (ritschwumm) [reply src/main/scala/scalaz/analytics/LocalAnalyticsModule.scala:56]: in one of my own projects i quite successfully used an offset variant of julian days (just counting solar days from some arbitrary day in the past, in my case 1.1.1970) for local dates instead - this has the great advantage of being completely independent of any calendar system.

**Comment by jdegoes** [comment]
@camjo Where did we end up on this? Just an Impala-like "TIMESTAMP" that's basically epoch time?

**Comment by camjo** [comment]
The merged PR has both instant and LocalDate. After rereading, I think the suggested approach above was to remove `LocalDate` and have it as a specialisation of a timezone-less `Instant`. Timezones would be stored as separate fields from the looks of it. @idc101 - Care to raise another PR dropping `LocalDate` for now until we either decide we need it or implement it on top of Instant?

**Comment by ritschwumm** [comment]
timezone-less, meaning something like a "LocalInstant"? sounds dangerous to me... or do you mean "as an offset from a fixed, absolute point in time, oblivious of the existence of timezones"?

**Comment by camjo** [comment]
@ritschwumm The latter. I probably could have phrased that better :)

**Pull request merged**